src/mrack/providers/openstack.py

```diff
@@ -145,6 +147,114 @@ def _set_images(self, images):
+        """Check if name is a configured network type in provisioning config."""
+        network_type = self.network_pools.get(name)
+        return bool(network_type)
+
+    def _aggregate_networks(self, hosts):
+        """
+        Get how many host require each used network type.
+
+        Returns: dict where keys are network types and values are total count.
+        """
+        network_types = {}
+        for host in hosts:
+            # skip hosts which have low-level network names defined
+            # this can be extended to pick network type based on the network name
+            names = host.get("networks")
+            if names:
+                continue
+            network_type = host.get("network")
+            if not self._is_network_type(network_type):
+                continue
+
+            count = network_types.get(network_type, 0)
+            count += 1
+            network_types[network_type] = count
+
+        return network_types
+
+    def _pick_network(self, network_type, count):
+        """
+        Pick network based network type and needed amount.
+
+        Usable network with most IPs available is picked.
+        """
+        possible_networks = self.network_pools[network_type]
+        networks = [self.get_network(net) for net in possible_networks]
+        usable = []
+        for network in networks:
+            ips = self.get_ips(ref=network.get("id"))
+
+            available = ips["total_ips"] - ips["used_ips"]
+            logger.debug(f"Network: {network['name']}")
+            logger.debug(f"  total: {ips['total_ips']}")
+            logger.debug(f"  used: {ips['used_ips']}")
+            logger.debug(f"  available: {available}")
+            if available > count:
+                usable.append((network["name"], available))
+        if not usable:
+            logger.error(
+                f"{self.dsp_name}: Error: no usable network"
+                f" for {count} hosts with {network_type}"
+            )
+            return None
+
+        # sort networks by number of available IPs
+        usable = sorted(usable, key=lambda u: u[1])
+        logger.debug(f"{self.dsp_name}: Listing usable networks: {usable}")
+
+        # Do not always pick the best, but randomize from good ones to spread
+        # load for high number of parallel jobs. E.g. if running mrack 100 times
+        # where each needs 3-4 hosts and best network has 250 IPs then we risk
+        # to pass the check intially but later fail as the check was not done
+        # with all the others on mind (race-condition).
+
+        # Good == has at least 50% of IPs as the best.
+        best_pool = usable[-1]
+        satisfying = [n for n in usable if n[1] / best_pool[1] > 0.5]
+        logger.debug(
+            f"{self.dsp_name}: Picking randomly from network pools which satisfy "
+            f"requirement (size_of_pool/size_of_biggest > 0.5): {satisfying}"
+        )
+        chosen = random.choice(satisfying)
```

Comment:
what am I doing :dagger:

Reply:
I see this in log:
```
OpenStack: Preparing provider resources
OpenStack: Validating hosts definitions
OpenStack: Validating host: {
    "config_drive": true,
    "flavor": "medium",
    "image": "Fedora-Cloud-Base-33-latest",
    "key_name": "idm-jenkins",
    "name": "f33-cont-0.mrack.test",
    "network": "net_9"
}
OpenStack: f33-cont-0.mrack.test - OK
OpenStack: Validating hosts definitions
OpenStack: Validating host: {
    "config_drive": true,
    "flavor": "medium",
    "image": "Fedora-Cloud-Base-33-latest",
    "key_name": "idm-jenkins",
    "name": "f33-cont-0.mrack.test",
    "network": "net_9"
}
OpenStack: f33-cont-0.mrack.test - OK
OpenStack: Host definitions valid
```
Is not that because of having one more set lines here?:
https://github.com/neoave/mrack/blob/master/src/mrack/providers/provider.py#L226-L228

Reply:
Should I remove that?

Reply:
https://github.com/neoave/mrack/pull/117/commits/ebce5e95771dd02e2b511f192235ffd8cc438b21